src/libraries/System.Text.Json/docs/JsonSerializationSourceGeneration.md

```diff
@@ -0,0 +1,183 @@
+        }
+
+        ReadOnlySpan<byte> propertyName;
+        SerializableClass obj = new SerializableClass();
+
+        if (!ReadPropertyName(ref reader)) goto Done;
+        propertyName = reader.ValueSpan;
+
+        if (propertyName.SequenceEqual(_s_Property_Email.NameAsUtf8Bytes))
+        {
+            reader.Read();
+            _s_Property_Email.ReadValueAndSetMember(ref reader, ref readStack, obj);
+            if (!ReadPropertyName(ref reader)) goto Done;
+            propertyName = reader.ValueSpan;
+        }
+
+        if (propertyName.SequenceEqual(_s_Property_Password.NameAsUtf8Bytes))
+        {
+            reader.Read();
+            _s_Property_Name.ReadValueAndSetMember(ref reader, ref readStack, obj);
+            if (!ReadPropertyName(ref reader)) goto Done;
+            propertyName = reader.ValueSpan;
+        }
+
+        reader.Read();
+
+    Done:
+        if (reader.TokenType != JsonTokenType.EndObject)
+        {
+            throw new JsonException("Could not deserialize object");
+        }
+
+        return obj;
+    }
+}
+
+```
+
+It is also important to notice that in case there are nested types within the root type we are recursing over, a new class with name ```FoundTypeNameSerializer``` will have to be created in order to completely serialize and deserialize.
+
+Even if the source generation fails, we can always fallback to the slower status quo by using ```Reflection```.
+
+#### Alternatives
+Some alternatives such as the use of interfaces for the functions mentioned above or the creation of individual JsonConverter for each type were talked about. However, due to performance and the direction we are taking with the initial prototype, we believe these are not necessary. After user feedback we can revisit this if needed. 
+
+### Generated Source Code Integration
+There are [discussions](https://gist.github.com/steveharter/d71cdfc25df53a8f60f1a3563d13cf0f) regarding integration of the approach mentioned above.
+
+The high level registration for the generated source code implies that the Json options class is modified by calling generated code where we use de/serialization API entry points for the extended class that auto-registers itself.
+
+For the most part the source code generation approach mentioned above solves this problem since the serialize and deserialize functions would live within the type class and once the serializer is initialized, the serializer can be used calling ```JsonSerializer.Serialize<Type>()``` or ```SourceGeneratedType.Serialize()```. In order to continue with this approach, an extension of the type class would be needed where the users would have to explicitly declare their types as partial classes for it to be extended by the source generator while the types they don’t own would be entirely created. 
+
+Even if this initialization isn't performed, we can always fallback to the slower status quo ```JsonSerializer``` methods.
```

Comment:
We discussed having a way to still look up the type's metadata within the context e.g. the context having a dictionary/if-else branches to return the right `JsonClassInfo`. I think this will be how reflection-based callers of the serializer e.g. ASP.NET can take advantage of this feature given that they may not have enough context to call the more specific overloads